File: TowerDefence.spritebuilder/Engine/TowerEngine.swift
//
//  TowerEngine.swift
//  TowerDefence
//
//  Created by koveres on 08.12.14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

import Foundation


public class TowerEngine {
 
    var towers:[Tower];
    var shells:[Shell];
 
    
    public  init( )
    {
        towers = [Tower]();
        shells = [Shell]();
    }
    
    public func getShells() -> ([Shell])
    {
        return shells;
    }
    
    public func   getTowers() -> ([Tower])
    {
        return towers;
    }
    
    
    
    func SetTower(  ID_Tower:TowerType,  position:CGPoint) -> (Bool) {
    
      var newTower:Tower = EngAndFabric.inst_.towerfabric.getTower(ID_Tower,position1: position,solderFinder1: EngAndFabric.inst_.solderFinder!);
        if(EngAndFabric.inst_.place!.canAddTower(newTower))
        {
        EngAndFabric.inst_.money!.Pay( newTower.getCost());
        towers.append(newTower);
        EngAndFabric.inst_.place!.addTower(newTower);
            return true;
        }
        return false;
    }
    
    func canSetTower( towerType:TowerType,   position:CGPoint) -> (Bool)
    {
    
        var newTower:Tower = EngAndFabric.inst_.towerfabric.getTower(towerType,position1: position,solderFinder1: EngAndFabric.inst_.solderFinder!);
    
        if(EngAndFabric.inst_.money!.canPay(newTower.getCost()))
        { return EngAndFabric.inst_.place!.canAddTower(newTower);}
        else
        {  return true;}
    }
    
    
    private func CheckShell() {
    
        for var i = (shells.count-1); i >= 0; --i
        {
       
            if( shells[i].isExploided())
            {
                    shells[i].FreeStackObject();
                shells.removeAtIndex( i );
            }
        }
    
    }
    
    func nextStep() {
    
    CheckShell();
        var shellCreated:Shell? = nil;
    for   tower in towers {
        shellCreated = tower.DoStep();
        if(shellCreated != nil) {
            shells.append(shellCreated!);
        }
    }
    
    for shell in shells
    {
        shell.DoStep(EngAndFabric.inst_.pathEng!,sold: EngAndFabric.inst_.soldeEng!);
    }
    
    }
    
    func soldTower(  ID:CInt) {
    
        for var i = 0; i < towers.count; ++i
    {
        if(towers[i].getIdObject() == ID)
        {
                EngAndFabric.inst_.money!.fillMoney(  towers[i].getSoldPrice()
                );
                towers[i].FreeStackObject();
                towers.removeAtIndex(i);
        }
    }
    }
    
}

File: test.spritebuilder/Source/TDEngine/Engine/TowerEngine.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Engine;

import Unit.Tower.Shell; 
import Unit.Tower.TargetFinder;
import Unit.Tower.Tower;
import Unit.Tower.TowerFabric;
import Utils.Point2D;
import internalObject.TowerInternal;
import internalObject.TowerInternal.TowerType;
 
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author vereshchagin
 */
class TowerEngine { 
    TowerFabric fabric;
    PlaceEngine place;
    List<Tower> towers;
    List<Shell> shells;
    TargetFinder solderFinder;
    MoneyEngine money;
    PathEngine pathEng;
    SolderEngine soldeEng;
     
    public TowerEngine(TowerFabric fabric,PlaceEngine place, SolderEngine unitE, MoneyEngine money,PathEngine pathEng)
    {
       this.fabric = fabric;
       this.place = place;
       solderFinder = new TargetFinder(unitE);
       towers = new ArrayList<Tower>();
       shells= new ArrayList<Shell>();
       this.money = money; 
       this.pathEng = pathEng;
       soldeEng = unitE;
    }
    public List<Shell> getShells()
    {
       return shells;
    }
    
    public List<Tower> getTowers()
    {
       return towers;
    }
    
    
    
    boolean SetTower(TowerInternal.TowerType ID_Tower,Point2D  position) throws Exception {
        
        Tower newTower = fabric.getTower(ID_Tower,position,solderFinder);
        if(place.canAddTower(newTower))
        {  
            money.Pay(newTower.getCost());
            towers.add(newTower);
            place.addTower(newTower);
            return true;
        } 
        return false;
    }

    boolean canSetTower(TowerType towerType, Point2D  position) throws Exception  
    {
        
        Tower newTower = fabric.getTower(towerType,position,solderFinder);
        
        if(money.canPay(newTower.getCost())) 
            return place.canAddTower(newTower);
        else
            return true;
    }

    
    private void CheckShell() {
      
        for(int i = shells.size()-1; i >=0; i-- ) 
        { 
            if( shells.get( i ).isExploided()) 
           {
               shells.get(i).FreeStackObject();
               shells.remove( i ); 
           }     
        }
 
    }
    
    void nextStep() {
         
        CheckShell();
        Shell shellCreated = null;
        for (Tower tower : towers) {
         
            shellCreated = tower.DoStep();
            if(shellCreated != null) {
                shells.add(shellCreated);
            }
        } 
        
        for (Shell shell : shells) 
        {
            shell.DoStep(this.pathEng,this.soldeEng);
        }
        
    }

    void soldTower(int ID) {
        for (int i = 0; i < towers.size(); i++) 
        {
            if(towers.get(i).getIdObject() == ID)
            {
               this.money.fillMoney((int) towers.get(i).getSoldPrice());
               towers.get(i).FreeStackObject();
               towers.remove(i); 
            }
        } 
        
        
    }
    
}
